// ts-express/dist/metadataGeneration/controllerGenerator.js
"use strict";
const ts = require("typescript");
const methodGenerator_1 = require("./methodGenerator");
class ControllerGenerator {
    constructor(node) {
        this.node = node;
        this.tagName = this.getControllerTagNameValue(node);
        this.jwtUserProperty = this.getControllerJWTValue(node);
    }
    IsValid() {
        return !!this.tagName || this.tagName === '';
    }
    Generate() {
        if (!this.node.parent) {
            throw new Error('Controller node doesn\'t have a valid parent source file.');
        }
        if (!this.node.name) {
            throw new Error('Controller node doesn\'t have a valid name.');
        }
        const sourceFile = this.node.parent.getSourceFile();
        return {
            jwtUserProperty: this.jwtUserProperty || '',
            location: sourceFile.fileName,
            methods: this.buildMethods(),
            name: this.node.name.text,
            tagName: this.tagName || ''
        };
    }
    buildMethods() {
        return this.node.members
            .filter(m => m.kind === ts.SyntaxKind.MethodDeclaration)
            .map((m) => new methodGenerator_1.MethodGenerator(m))
            .filter(generator => generator.IsValid())
            .map(generator => generator.Generate());
    }
    getControllerJWTValue(node) {
        return this.getControllerDecoratorValue(node, 'JWT', 'user');
    }
    getControllerTagNameValue(node) {
        return this.getControllerDecoratorValue(node, 'Controller', '');
    }
    getControllerDecoratorValue(node, decoratorName, defaultValue) {
        if (!node.decorators) {
            return undefined;
        }
        const matchedAttributes = node.decorators
            .map(d => d.expression)
            .filter(expression => {
            const subExpression = expression.expression;
            return subExpression.text === decoratorName;
        });
        if (!matchedAttributes.length) {
            return undefined;
        }
        if (matchedAttributes.length > 1) {
            throw new Error('A controller can only have a single "`decoratorName`" decorator.');
        }
        const value = matchedAttributes[0].arguments[0];
        return value ? value.text : defaultValue;
    }
}
exports.ControllerGenerator = ControllerGenerator;
